mpaste.1
.\" vi: tw=100
.Dd 24 January, 2022
.Dt MPASTE 1
.Os \*(Px
.Sh NAME
.Nm mpaste
.Nd a simple and minimal paste server
.Sh SYNOPSIS
.Nm
.Op Fl c Ar counter
.Op Fl f Ar file_dir
.Op Fl i Ar index
.Op Fl u Ar users
.Ar domain
.Ar port
.Sh DESCRIPTION
.Nm
is a minimal paste server for hosting plaintext data.
The paste server has support for file uploads, syntax highlighting, a customizable homepage,
customizable tab width, and a password protected mode where only users with valid API keys can
upload pastes.
For the simplest example of a working paste server, simply run
.Nm
and provide it with a
.Ar domain
and
.Ar port .
The provided domain is not super important, it is just used in the message sent back to the client
after a successful paste so that they have a direct link to click on to go to their paste.
The port on the otherhand does matter, it is the port on which the server will listen.
.Pp
Once the server is running you can POST a file to the server by sending a form with the name
.Dq file .
Here is an example of POSTing a file with
.Xr curl 1 :
.Pp
.Dl $ curl -X POST -F \(aqfile=@foo.txt\(aq domain.com
.Pp
After a successful POST the server will respond with a URI to the post in the form
.Dq domain.com/ID
where
.Dq ID
is a number which increments with each paste.
When viewed, the paste will be displayed as unformatted plaintext.
If you would like syntax highlighting simply append the appropiate file extension to the URI.
For example, to syntax highlight C code with paste ID 5, go to
.Dq domain.com/5.c .
.Pp
By default when syntax highlighting the server displays tab characters with a width of 8 columns. If
you would like to customize the width of a tab you can set the
.Dq tabs
query parameter in the URI.
For example, to view a Python file with a tab width of 4 columns, go to
.Dq domain.com/1.py?tabs=4 .
.Pp
If you would like to protect the server by requiring all users to have an API key, simply set the
.Ev MPASTE_SECRET
environment variable.
With this secret set, you can generate a JWT token encoded with that same secret, and with the
playload
.Dq name=USERS NAME .
For example, one might have the payload
.Dq name=Johnny Appleseed .
This name is then looked up in the
.Pa users
file.
If the name is found in that file, the POST is allowed, otherwise it is rejected.
You can specify the
.Pa users
file with the
.Fl u
flag.
An example file might look like this:
.Pp
.Bd -literal -offset indent
Johnny Appleseed
John Doe
Hunter
.Ed
.Pp
As a user if you want to authenticate yourself you must send your JWT token in an authorization
header.
As an example using
.Xr curl 1 :
.Pp
.Dl $ curl -X POST -H \(aqAuthorization: YOUR.JWT.TOKEN\(aq -F \(aqfile=@foo.txt\(aq domain.com
.Pp
Finally, you may want to display content on the paste servers homepage.
This is easy and can be done by creating a
.Pa index.html
in the current working directory.
If you would like to specify a different file, you can use the
.Fl i Ar index
flag.
.Sh OPTIONS
.Bl -tag -width Ds
.It Fl c Ar counter
Specify the path to a file to use as a counter.
This file will hold the number of the ID that will be assigned to the next paste.
If the given file does not exist, it will be created on the next successful paste.
If this flag is not specified then it will default to
.Pa counter .
.It Fl f Ar file_dir
Specify a directory in which to store the pastes that users POST to the server.
If the given folder does not exist, then it will be created.
If this flag is not specified then it will default to
.Pa files/ .
.It Fl i Ar index
Specify a file to serve on the servers root
.Pq Pa /
page.
If this flag is not specified then it will default to
.Pa index.html .
.It Fl u Ar users
Specify a file to store authorized users in.
This file must be created by the user and must contain a newline seperated list of authorized users
as shown in the
.Sx DESCRIPTION
section of this manual.
If this flag is specified and the
.Ev MPASTE_SECRET
environment variable is not set, it will have no effect.
If the environment is set and this file does not exist, then no users will be allowed to POST.
.El
.Sh FILES
.Bl -tag -width Ds
.It Pa counter
This is where the ID of the next paste is stored.
.It Pa index.html
This is the default file that the
.Nm
server will attempt to serve on the root
.Pq Pa /
page.
.It Pa users
This is a newline seperated list of authenticated users.
.El
.Sh ENVIRONMENT
.Bl -tag -width Ds
.It Ev MPASTE_SECRET
This is the secret key used to encode and decode the JWT tokens used when authenticating users.
Under no circumstances should you share this token with anybody.
If not set, anyone will be able to POST their pastes to the server.
.El
.Sh EXIT STATUS
.Ex -std
.Sh SEE ALSO
.Xr curl 1 ,
.Xr nginx 1
.Sh AUTHORS
.An Thomas Voss Aq Mt mail@thomasvoss.com
.Sh SECURITY CONSIDERATIONS
If deployed on a public network
.Pq or even on a private one
you should take the following
.Pq non-exhaustive
list of scenarios into consideration:
.Bl -dash
.It
Users uploading exessively large files. You can consider using tools such as
.Xr nginx 1
to control the maximum allowed file upload size.
.It
Users uploading exessively many files.
.It
Users uploading non-plaintext files. On certain browsers this may prompt a user to download the
hosted content, which is a potential attack vector.
.El


mpaste.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"io/ioutil"
	"net/http"
	"os"
	"path"
	"strconv"
	"strings"
	"sync"

	"github.com/Mango0x45/getgopt"
	"github.com/alecthomas/chroma/formatters/html"
	"github.com/alecthomas/chroma/lexers"
	"github.com/alecthomas/chroma/styles"
	"github.com/dgrijalva/jwt-go"
)

const (
	urlHomepage = iota
	urlInvalid
	urlSyntax
	urlValid
)

var (
	counterFile string
	counter     int
	domain      string
	filePrefix  string
	indexFile   string
	mutex       sync.Mutex
	secretKey   = os.Getenv("MPASTE_SECRET")
	style       = styles.Get("monokai")
	userFile    string
)

func usage() {
	fmt.Fprintf(os.Stderr,
		"Usage: %s [-c file] [-f directory] [-i file] [-u file] domain port\n",
		os.Args[0])
	os.Exit(1)
}

func die(e interface{}) {
	fmt.Fprintln(os.Stderr, e)
	os.Exit(1)
}

func writeHeader(w http.ResponseWriter, h int, s string) {
	w.WriteHeader(h)
	if s == "" {
		fmt.Fprintln(w, http.StatusText(h))
	} else {
		fmt.Fprintln(w, s)
	}
}

func removeExt(s string) string {
	return strings.TrimSuffix(s, path.Ext(s))
}

func allowedUser(name string) bool {
	mutex.Lock()
	defer mutex.Unlock()

	if _, err := os.Stat(userFile); os.IsNotExist(err) {
		return false
	}

	file, err := os.Open(userFile)
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		return false
	}

	defer file.Close()

	scanner := bufio.NewScanner(file)
	scanner.Split(bufio.ScanLines)

	for scanner.Scan() {
		if scanner.Text() == name {
			return true
		}
	}

	return false
}

func validateToken(r *http.Request) bool {
	token, _ := jwt.Parse(r.Header.Get("Authorization"), func(t *jwt.Token) (interface{}, error) {
		if _, ok := t.Method.(*jwt.SigningMethodHMAC); !ok {
			return nil, nil
		}
		return []byte(secretKey), nil
	})

	if token == nil {
		return false
	}

	claims, ok := token.Claims.(jwt.MapClaims)

	if !(ok && token.Valid) {
		return false
	}

	if userFile == "" {
		return true
	}

	return allowedUser(claims["name"].(string))
}

func isValidUrl(s string) int {
	var i int
	var c rune
	for i, c = range s {
		if c == '.' && i > 0 {
			return urlSyntax
		} else if c < '0' || c > '9' {
			return urlInvalid
		}
	}

	if c != 0 {
		return urlValid
	}
	return urlHomepage
}

func syntaxHighlighting(w http.ResponseWriter, r *http.Request) {
	lexer := lexers.Match(r.URL.Path[1:])
	if lexer == nil {
		http.ServeFile(w, r, filePrefix+r.URL.Path[1:])
		return
	}

	data, err := ioutil.ReadFile(filePrefix + removeExt(r.URL.Path[1:]))
	if err != nil {
		w.Header().Set("Content-Type", "text/plain; charset=utf-8")
		writeHeader(w, http.StatusNotFound, "")
		return
	}

	iterator, err := lexer.Tokenise(nil, string(data))
	if err != nil {
		writeHeader(w, http.StatusInternalServerError, "Failed to tokenize output")
		return
	}

	tw, err := strconv.Atoi(r.URL.Query().Get("tabs"))
	if err != nil {
		tw = 8
	}
	formatter := html.New(html.Standalone(true), html.WithClasses(true),
		html.WithLineNumbers(true), html.LineNumbersInTable(true), html.TabWidth(tw))
	if err := formatter.Format(w, style, iterator); err != nil {
		writeHeader(w, http.StatusInternalServerError, "Failed to format output")
		return
	}
}

func endpoint(w http.ResponseWriter, r *http.Request) {
	switch r.Method {
	case http.MethodGet:
		switch isValidUrl(r.URL.Path[1:]) {
		case urlHomepage:
			http.ServeFile(w, r, indexFile)
		case urlInvalid:
			writeHeader(w, http.StatusNotFound, "")
			return
		case urlSyntax:
			w.Header().Set("Content-Type", "text/html; charset=utf-8")
			syntaxHighlighting(w, r)
		case urlValid:
			w.Header().Set("Content-Type", "text/plain; charset=utf-8")
			http.ServeFile(w, r, filePrefix+r.URL.Path[1:])
		}
	case http.MethodPost:
		if secretKey != "" && !validateToken(r) {
			writeHeader(w, http.StatusForbidden, "")
			return
		}

		file, _, err := r.FormFile("data")
		defer file.Close()
		if err != nil {
			writeHeader(w, http.StatusInternalServerError, "Failed to parse form")
			return
		}

		mutex.Lock()
		defer mutex.Unlock()

		fname := filePrefix + strconv.Itoa(counter)
		nfile, err := os.Create(fname)
		defer nfile.Close()
		if err != nil {
			writeHeader(w, http.StatusInternalServerError, "Failed to create file")
			return
		}

		if _, err = io.Copy(nfile, file); err != nil {
			writeHeader(w, http.StatusInternalServerError, "Failed to write file")
			return
		}

		if err = os.WriteFile(counterFile, []byte(strconv.Itoa(counter+1)), 0644); err != nil {
			writeHeader(w, http.StatusInternalServerError, "Failed to update counter")
			return
		}

		w.WriteHeader(http.StatusOK)
		fmt.Fprintf(w, domain+"/%d\n", counter)

		counter++
	default:
		writeHeader(w, http.StatusMethodNotAllowed, "Only GET and POST requests are supported")
	}
}

func main() {
	for opt := byte(0); getgopt.Getopt(len(os.Args), os.Args, ":c:f:i:u:", &opt); {
		switch opt {
		case 'c':
			counterFile = getgopt.Optarg
		case 'f':
			filePrefix = getgopt.Optarg
		case 'i':
			indexFile = getgopt.Optarg
		case 'u':
			userFile = getgopt.Optarg
		default:
			usage()
		}
	}

	argv := os.Args[getgopt.Optind:]
	if len(argv) != 2 {
		usage()
	}
	domain = argv[0]
	port := argv[1]

	if filePrefix == "" {
		filePrefix = "files/"
	} else if filePrefix[len(filePrefix)-1] != '/' {
		filePrefix += "/"
	}

	if counterFile == "" {
		counterFile = "counter"
	}

	if indexFile == "" {
		indexFile = "index.html"
	}

	if _, err := os.Stat(indexFile); os.IsNotExist(err) {
		die(err)
	}

	if _, err := os.Stat(filePrefix); os.IsNotExist(err) {
		if err = os.MkdirAll(filePrefix, 0755); err != nil {
			die(err)
		}
	}

	if _, err := os.Stat(counterFile); os.IsNotExist(err) {
		counter = 0
	} else {
		data, err := ioutil.ReadFile(counterFile)
		if err != nil {
			die(err)
		}
		counter, _ = strconv.Atoi(string(data))
	}

	http.HandleFunc("/", endpoint)
	die(http.ListenAndServe(":"+port, nil))
}
